bookwyrm/views/list.py

```diff
@@ -210,11 +284,103 @@ def add_book(request):
 @require_POST
 def remove_book(request, list_id):
     """ put a book on a list """
-    book_list = get_object_or_404(models.List, id=list_id)
-    item = get_object_or_404(models.ListItem, id=request.POST.get("item"))
+    with transaction.atomic():
+        book_list = get_object_or_404(models.List, id=list_id)
+        item = get_object_or_404(models.ListItem, id=request.POST.get("item"))
 
-    if not book_list.user == request.user and not item.user == request.user:
-        return HttpResponseNotFound()
+        if not book_list.user == request.user and not item.user == request.user:
+            return HttpResponseNotFound()
 
-    item.delete()
+        deleted_order = item.order
+        item.delete()
+    normalize_book_list_ordering(book_list.id, start=deleted_order)
     return redirect("list", list_id)
+
+
+@require_POST
+def set_book_position(request, list_item_id):
+    """
+    Action for when the list user manually specifies a list position, takes special care with the unique ordering per list
+    """
+    with transaction.atomic():
+        list_item = get_object_or_404(models.ListItem, id=list_item_id)
+        position = request.POST.get("position")
+        try:
+            int_position = int(request.POST.get("position"))
+        except ValueError:
+            return HttpResponseBadRequest(
+                "bad value for position. should be an integer"
+            )
+
+        if int_position < 1:
+            return HttpResponseBadRequest("position cannot be less than 1")
+
+        book_list = list_item.book_list
+        order_max = book_list.listitem_set.aggregate(Max("order"))["order__max"]
+
+        if int_position > order_max:
+            int_position = order_max
+
+        if request.user not in (book_list.user, list_item.user):
+            return HttpResponseNotFound()
+
+        original_order = list_item.order
+        if original_order == int_position:
+            return
+        elif original_order > int_position:
+            list_item.order = -1
+            list_item.save()
+            increment_order_in_reverse(book_list.id, int_position, original_order)
+        else:
+            list_item.order = -1
+            list_item.save()
+            decrement_order(book_list.id, original_order, int_position)
+
+        list_item.order = int_position
+        list_item.save()
+
+    return redirect("list", book_list.id)
+
+
+@transaction.atomic
+def increment_order_in_reverse(book_list_id, start, end):
+    print(f"increment_order_in_reverse, {book_list_id=}, {start=}, {end=}")
```

Comment:
I believe these print statements should get removed

Reply:
Thanks. They should be gone now.